File: frontend/src/api/ra.ts
/**
 * RA 相关 API
 */
import apiClient from './index'

export type TopSort = 'claim' | 'ra'

// 筛选选项
export interface RaOptions {
  time_range: {
    min_month: string
    max_month: string
  }
  odms: string[]
  segments: string[]
  models: string[]
}

// 趋势数据点
export interface TrendPoint {
  month: string
  ra: number
  ra_claim?: number
  ra_mm?: number
}

// Top Model
export interface TopModel {
  rank: number
  model: string
  ra: number
  ra_claim: number
  ra_mm: number
  top_issues?: TopIssue[]
}

// Top ODM
export interface TopOdm {
  rank: number
  odm: string
  ra: number
  ra_claim: number
  ra_mm: number
}

// Top Issue
export interface TopIssue {
  rank: number
  issue: string
  count: number
  share?: number
}

// 月度明细
export interface MonthlyTopModels {
  month: string
  items: TopModel[]
}

export interface MonthlyTopOdms {
  month: string
  items: TopOdm[]
}

export interface MonthlyTopIssues {
  month: string
  items: TopIssue[]
}

// 饼图数据
export interface OdmPieItem {
  odm: string
  ra: number
  share: number
  ra_claim?: number
  ra_mm?: number
}

export interface SegmentPieItem {
  segment: string
  ra: number
  share: number
  ra_claim?: number
  ra_mm?: number
}

export interface ModelPieItem {
  model: string
  ra: number
  share: number
  ra_claim?: number
  ra_mm?: number
}

// ODM 分析卡片
export interface OdmCard {
  odm: string
  trend: TrendPoint[]
  top_models: TopModel[]
  monthly_top_models?: MonthlyTopModels[]
  ai_summary: string
}

// Segment 分析卡片
export interface SegmentCard {
  segment: string
  trend: TrendPoint[]
  top_odms: TopOdm[]
  top_models: TopModel[]
  monthly_top_odms?: MonthlyTopOdms[]
  monthly_top_models?: MonthlyTopModels[]
  ai_summary: string
}

// Model 分析卡片
export interface ModelCard {
  model: string
  trend: TrendPoint[]
  top_issues: TopIssue[]
  monthly_top_issues?: MonthlyTopIssues[]
  ai_summary: string
}

// Summary 数据 (Block D)
export interface OdmSummary {
  odm_pie: OdmPieItem[]
}

export interface SegmentSummary {
  segment_pie: SegmentPieItem[]
}

export interface ModelSummary {
  model_pie: ModelPieItem[]
}

// 分析元数据
export interface AnalyzeMeta {
  data_as_of: string
  time_range: {
    start_month: string
    end_month: string
  }
}

// ODM 分析响应
export interface OdmAnalyzeResponse {
  meta: AnalyzeMeta
  cards: OdmCard[]
  summary?: OdmSummary
}

// Segment 分析响应
export interface SegmentAnalyzeResponse {
  meta: AnalyzeMeta
  cards: SegmentCard[]
  summary?: SegmentSummary
}

// Model 分析响应
export interface ModelAnalyzeResponse {
  meta: AnalyzeMeta
  cards: ModelCard[]
  summary?: ModelSummary
}

/**
 * 获取 RA 筛选选项（支持任意组合过滤）
 */
export async function getRaOptions(filters?: {
  segments?: string[]
  odms?: string[]
  models?: string[]
}): Promise<RaOptions> {
  const params: Record<string, string> = {}
  if (filters?.segments?.length) params.segments = filters.segments.join(',')
  if (filters?.odms?.length) params.odms = filters.odms.join(',')
  if (filters?.models?.length) params.models = filters.models.join(',')
  const response = await apiClient.get('/ra/options', { 
    params: Object.keys(params).length ? params : undefined 
  })
  return response.data
}

// 兼容旧接口
export const getRaOdmOptions = (segments?: string[]) => 
  getRaOptions(segments?.length ? { segments } : undefined)

export const getRaSegmentOptions = (odms?: string[]) => 
  getRaOptions(odms?.length ? { odms } : undefined)

export const getRaModelOptions = (segments?: string[], odms?: string[]) => 
  getRaOptions({ segments, odms })

/**
 * RA ODM 分析
 */
export async function analyzeRaOdm(params: {
  start_month: string
  end_month: string
  odms: string[]
  segments?: string[]
  models?: string[]
  top_model_sort?: TopSort
}): Promise<OdmAnalyzeResponse> {
  const request = {
    time_range: {
      start_month: params.start_month,
      end_month: params.end_month
    },
    filters: {
      odms: params.odms,
      segments: params.segments,
      models: params.models
    },
    view: {
      top_model_sort: params.top_model_sort
    }
  }
  const response = await apiClient.post('/ra/odm-analysis/analyze', request)
  return response.data
}

/**
 * RA Segment 分析
 */
export async function analyzeRaSegment(params: {
  start_month: string
  end_month: string
  segments: string[]
  odms?: string[]
  models?: string[]
  top_odm_sort?: TopSort
  top_model_sort?: TopSort
}): Promise<SegmentAnalyzeResponse> {
  const request = {
    time_range: {
      start_month: params.start_month,
      end_month: params.end_month
    },
    filters: {
      segments: params.segments,
      odms: params.odms,
      models: params.models
    },
    view: {
      top_odm_sort: params.top_odm_sort,
      top_model_sort: params.top_model_sort
    }
  }
  const response = await apiClient.post('/ra/segment-analysis/analyze', request)
  return response.data
}

/**
 * RA Model 分析
 */
export async function analyzeRaModel(params: {
  start_month: string
  end_month: string
  models: string[]
  segments?: string[]
  odms?: string[]
}): Promise<ModelAnalyzeResponse> {
  const request = {
    time_range: {
      start_month: params.start_month,
      end_month: params.end_month
    },
    filters: {
      models: params.models,
      segments: params.segments,
      odms: params.odms
    }
  }
  const response = await apiClient.post('/ra/model-analysis/analyze', request)
  return response.data
}

// ==================== Report Download ====================

export interface ReportDownloadResult {
  blob: Blob
  filename?: string
}

function _getFilename(headers: Record<string, string>): string | undefined {
  const cd = headers['content-disposition']
  if (!cd) return undefined
  const m = cd.match(/filename\*=UTF-8''([^;]+)/i)
  if (m?.[1]) return decodeURIComponent(m[1])
  const m2 = cd.match(/filename="?([^";]+)"?/i)
  return m2?.[1]
}

export async function downloadRaModelReport(params: {
  start_month: string
  end_month: string
  models: string[]
  segments?: string[]
  odms?: string[]
  tgt?: number
}): Promise<ReportDownloadResult> {
  const request = {
    time_range: { start_month: params.start_month, end_month: params.end_month },
    filters: { models: params.models, segments: params.segments, odms: params.odms },
    tgt: params.tgt,
  }
  const response = await apiClient.post('/ra/report/model', request, {
    responseType: 'blob', timeout: 120000,
  })
  return { blob: response.data, filename: _getFilename(response.headers as any) }
}

export async function downloadRaOdmReport(params: {
  start_month: string
  end_month: string
  odms: string[]
  segments?: string[]
  models?: string[]
  top_model_sort?: TopSort
  tgt?: number
}): Promise<ReportDownloadResult> {
  const request = {
    time_range: { start_month: params.start_month, end_month: params.end_month },
    filters: { odms: params.odms, segments: params.segments, models: params.models },
    view: { top_model_sort: params.top_model_sort },
    tgt: params.tgt,
  }
  const response = await apiClient.post('/ra/report/odm', request, {
    responseType: 'blob', timeout: 120000,
  })
  return { blob: response.data, filename: _getFilename(response.headers as any) }
}

export async function downloadRaSegmentReport(params: {
  start_month: string
  end_month: string
  segments: string[]
  odms?: string[]
  models?: string[]
  top_odm_sort?: TopSort
  top_model_sort?: TopSort
  tgt?: number
}): Promise<ReportDownloadResult> {
  const request = {
    time_range: { start_month: params.start_month, end_month: params.end_month },
    filters: { segments: params.segments, odms: params.odms, models: params.models },
    view: { top_odm_sort: params.top_odm_sort, top_model_sort: params.top_model_sort },
    tgt: params.tgt,
  }
  const response = await apiClient.post('/ra/report/segment', request, {
    responseType: 'blob', timeout: 120000,
  })
  return { blob: response.data, filename: _getFilename(response.headers as any) }
}


// Issue 明细行
export interface IssueDetailRow {
  model: string
  fault_category: string
  problem_descr_by_tech?: string
  claim_nbr: string
  claim_month: string
  plant?: string
}

// Issue 明细响应
export interface IssueDetailResponse {
  total: number
  page: number
  page_size: number
  items: IssueDetailRow[]
}

/**
 * RA Model Issue 明细
 */
export async function getRaModelIssueDetails(params: {
  start_month: string
  end_month: string
  model: string
  issue: string
  month?: string
  segments?: string[]
  odms?: string[]
  page?: number
  page_size?: number
}): Promise<IssueDetailResponse> {
  const request = {
    time_range: {
      start_month: params.start_month,
      end_month: params.end_month
    },
    filters: {
      model: params.model,
      issue: params.issue,
      month: params.month,
      segments: params.segments,
      odms: params.odms
    },
    pagination: {
      page: params.page ?? 1,
      page_size: params.page_size ?? 10
    }
  }
  const response = await apiClient.post('/ra/model-analysis/issue-details', request)
  return response.data
}

File: backend/app/schemas/ra.py
"""
RA相关Schema定义
"""
from typing import Optional, List, Literal
from pydantic import BaseModel
from app.schemas.common import TimeRange, OptionsTimeRange


# ==================== Options ====================

class RaOptionsData(BaseModel):
    """RA筛选项数据"""
    month_min: str
    month_max: str
    data_as_of: str
    time_range: OptionsTimeRange
    segments: List[str]
    odms: List[str]
    models: List[str]


class RaOptionsResponse(BaseModel):
    """RA Options API响应"""
    code: int = 0
    message: str = "success"
    data: Optional[RaOptionsData] = None


# ==================== ODM Analyze ====================

class RaOdmFilters(BaseModel):
    """RA ODM分析筛选条件"""
    odms: List[str]
    segments: Optional[List[str]] = None
    models: Optional[List[str]] = None


class RaOdmViewConfig(BaseModel):
    """RA ODM分析视图配置"""
    trend_months: int = 6
    top_model_n: int = 10
    top_model_sort: Literal["claim", "ra"] = "claim"


class RaOdmAnalyzeRequest(BaseModel):
    """RA ODM分析请求"""
    time_range: TimeRange
    filters: RaOdmFilters
    view: Optional[RaOdmViewConfig] = None


class RaTrendPoint(BaseModel):
    """RA趋势数据点"""
    month: str
    ra: float


class RaTopModelRow(BaseModel):
    """Top Model数据行"""
    rank: int
    model: str
    ra: float
    ra_claim: Optional[int] = None
    ra_mm: Optional[int] = None


class RaMonthlyTopModels(BaseModel):
    """月度Top Model明细"""
    month: str
    items: List[RaTopModelRow]


class RaOdmPieRow(BaseModel):
    """ODM饼图数据行"""
    odm: str
    ra: float
    share: float
    ra_claim: Optional[int] = None
    ra_mm: Optional[int] = None


class RaOdmCard(BaseModel):
    """RA ODM分析卡片"""
    odm: str
    trend: List[RaTrendPoint]
    top_models: List[RaTopModelRow]  # 汇总数据
    monthly_top_models: Optional[List[RaMonthlyTopModels]] = None  # 月度明细
    ai_summary: str = ""


class RaOdmSummary(BaseModel):
    """RA ODM汇总数据(Block D)"""
    odm_pie: List[RaOdmPieRow]


class RaAnalyzeMeta(BaseModel):
    """RA分析元数据"""
    data_as_of: str
    time_range: TimeRange


class RaOdmAnalyzeData(BaseModel):
    """RA ODM分析响应数据"""
    meta: RaAnalyzeMeta
    summary: Optional[RaOdmSummary] = None
    cards: List[RaOdmCard]


class RaOdmAnalyzeResponse(BaseModel):
    """RA ODM分析响应"""
    code: int = 0
    message: str = "success"
    data: Optional[RaOdmAnalyzeData] = None


# ==================== Segment Analyze ====================

class RaSegmentFilters(BaseModel):
    """RA Segment分析筛选条件"""
    segments: List[str]
    odms: Optional[List[str]] = None
    models: Optional[List[str]] = None


class RaSegmentViewConfig(BaseModel):
    """RA Segment分析视图配置"""
    trend_months: int = 6
    top_n: int = 10
    top_odm_sort: Literal["claim", "ra"] = "claim"
    top_model_sort: Literal["claim", "ra"] = "claim"


class RaSegmentAnalyzeRequest(BaseModel):
    """RA Segment分析请求"""
    time_range: TimeRange
    filters: RaSegmentFilters
    view: Optional[RaSegmentViewConfig] = None


class RaTopOdmRow(BaseModel):
    """Top ODM数据行"""
    rank: int
    odm: str
    ra: float
    ra_claim: Optional[int] = None
    ra_mm: Optional[int] = None


class RaMonthlyTopOdms(BaseModel):
    """月度Top ODM明细"""
    month: str
    items: List[RaTopOdmRow]


class RaSegmentPieRow(BaseModel):
    """Segment饼图数据行"""
    segment: str
    ra: float
    share: float
    ra_claim: Optional[int] = None
    ra_mm: Optional[int] = None


class RaSegmentCard(BaseModel):
    """RA Segment分析卡片"""
    segment: str
    trend: List[RaTrendPoint]
    top_odms: List[RaTopOdmRow]  # 汇总数据
    top_models: List[RaTopModelRow]  # 汇总数据
    monthly_top_odms: Optional[List[RaMonthlyTopOdms]] = None  # 月度明细
    monthly_top_models: Optional[List[RaMonthlyTopModels]] = None  # 月度明细
    ai_summary: str = ""


class RaSegmentSummary(BaseModel):
    """RA Segment汇总数据(Block D)"""
    segment_pie: List[RaSegmentPieRow]


class RaSegmentAnalyzeData(BaseModel):
    """RA Segment分析响应数据"""
    meta: RaAnalyzeMeta
    summary: Optional[RaSegmentSummary] = None
    cards: List[RaSegmentCard]


class RaSegmentAnalyzeResponse(BaseModel):
    """RA Segment分析响应"""
    code: int = 0
    message: str = "success"
    data: Optional[RaSegmentAnalyzeData] = None


# ==================== Model Analyze ====================

class RaModelFilters(BaseModel):
    """RA Model分析筛选条件"""
    models: List[str]
    segments: Optional[List[str]] = None
    segment: Optional[str] = None
    odms: Optional[List[str]] = None


class RaModelViewConfig(BaseModel):
    """RA Model分析视图配置"""
    top_issue_n: int = 10


class RaModelAnalyzeRequest(BaseModel):
    """RA Model分析请求"""
    time_range: TimeRange
    filters: RaModelFilters
    view: Optional[RaModelViewConfig] = None


class RaTopIssueRow(BaseModel):
    """Top Issue数据行"""
    rank: int
    issue: str
    count: int
    share: Optional[float] = None


class RaMonthlyTopIssues(BaseModel):
    """月度Top Issue明细"""
    month: str
    items: List[RaTopIssueRow]


class RaModelPieRow(BaseModel):
    """Model饼图数据行"""
    model: str
    ra: float
    share: float
    ra_claim: Optional[int] = None
    ra_mm: Optional[int] = None


class RaModelCard(BaseModel):
    """RA Model分析卡片"""
    model: str
    trend: List[RaTrendPoint]
    top_issues: List[RaTopIssueRow]  # 汇总数据
    monthly_top_issues: Optional[List[RaMonthlyTopIssues]] = None  # 月度明细
    ai_summary: str = ""


class RaModelSummary(BaseModel):
    """RA Model汇总数据(Block D)"""
    model_pie: List[RaModelPieRow]


class RaModelAnalyzeData(BaseModel):
    """RA Model分析响应数据"""
    meta: RaAnalyzeMeta
    summary: Optional[RaModelSummary] = None
    cards: List[RaModelCard]


class RaModelAnalyzeResponse(BaseModel):
    """RA Model分析响应"""
    code: int = 0
    message: str = "success"
    data: Optional[RaModelAnalyzeData] = None


